File: leetcode/001-two-sum/Solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i]) && i != map.get(nums[i])) {
                return new int[] {i, map.get(nums[i])};
            }
            map.put(target - nums[i], i);
        }

        return null;
    }

    public static void main(String[] args) {
        Input[] inputs = {
            new Input(new int[]{ 2, 7, 11, 15 }, 9),
        };

        for (int i = 0; i < inputs.length; i++) {
            int[] nums = inputs[i].nums;
            int target = inputs[i].target;

            System.out.print("\n Given nums: ");
            System.out.print(Arrays.toString(nums));
            System.out.printf(", target = %d\n", target);

            Solution solution = new Solution();
            int[] pairs = solution.twoSum(nums, target);
            System.out.print(" Output: ");
            System.out.printf("%s\n", (pairs == null) ? "Not found" : Arrays.toString(pairs));
        }
    }

    static class Input {
        int[] nums;
        int target;

        Input(int[] nums, int target) {
            this.nums = nums;
            this.target = target;
        }
    }
}

File: leetcode/001-two-sum/solution.py
# -*- coding: utf-8 -*-


class Solution:

    def twoSum(self, nums: list, target: int) -> list:
        cache = {}
        for i, num in enumerate(nums):
            if num in cache and i != cache[num]:
                return [i, cache[num]]
            cache[target - num] = i
        return []

    def twoSum2(self, nums: list, target: int) -> list:
        idx_pos, val_pos = 0, 1
        objs = [(i, num) for i, num in enumerate(nums)]
        objs.sort(key=lambda x: x[val_pos])
        low, high = 0, len(nums) - 1
        while low < high:
            diff = target - objs[low][val_pos]
            if diff > objs[high][val_pos]:
                while low + 1 < high and objs[low][val_pos] == objs[low + 1][val_pos]: low += 1
                low += 1
            elif diff < objs[high][val_pos]:
                while high - 1 >= low and objs[high][val_pos] == objs[high - 1][val_pos]: high -= 1
                high -= 1
            else:
                return [objs[low][idx_pos], objs[high][idx_pos]]
        return []


if __name__ == '__main__':
    inputs = [
        ([2, 7, 11, 15], 9)
    ]

    for nums, target in inputs:
        print(' Given nums: ', nums, ', target = ', target, sep='')
        solution = Solution()
        pos = solution.twoSum(nums, target)
        print(' Output: ', pos, sep='')
        pos = solution.twoSum2(nums, target)
        print(' Output: ', pos, sep='')
